supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala

```diff
@@ -0,0 +1,142 @@
+      jsonStringIo.flatMap(jsonStr => IO.fromEither(decode[ManifestFile](jsonStr)))
+    }
+  }
+
+  private def getReferenceFileToValidatedGcsPathMap(referenceFiles: Set[ReferenceFile]): IO[Map[ReferenceFile, ValidFullGcsPath]] = {
+    val filesAndValidatedPaths = referenceFiles.map {
+      referenceFile => (referenceFile, GcsPathBuilder.validateGcsPath(s"gs://${referenceFile.path}"))
+    }.toMap
+
+    val filesWithValidPaths = filesAndValidatedPaths.collect {
+      case (referenceFile, validPath: ValidFullGcsPath) => (referenceFile, validPath)
+    }
+    val filesWithInvalidPaths = filesAndValidatedPaths.collect {
+      case (referenceFile, invalidPath: InvalidFullGcsPath) => (referenceFile, invalidPath)
+    }
+
+    if (filesWithInvalidPaths.nonEmpty) {
+      IO.raiseError(new InvalidGcsPathsInManifestFileException(filesWithInvalidPaths.keySet.map(_.path).toList))
+    } else {
+      IO.pure(filesWithValidPaths)
+    }
+  }
+
+  protected def bulkValidateCrc32cs(gcsClient: Storage,
+                          filesWithValidPathsIo: IO[Map[ReferenceFile, ValidFullGcsPath]]): IO[Map[ReferenceFile, Boolean]] = {
+    val gcsBatch = gcsClient.batch()
+    val filesAndBlobResultsIo = filesWithValidPathsIo.map { filesWithValidPaths =>
+      val filesAndBlobResults = filesWithValidPaths map {
+        case (referenceFile, ValidFullGcsPath(bucket, path)) =>
+          val blobGetResult = gcsBatch.get(BlobId.of(bucket, path.substring(1)), BlobGetOption.fields(BlobField.CRC32C))
+          (referenceFile, blobGetResult)
+      }
+      gcsBatch.submit()
+      filesAndBlobResults
+    }
+
+    filesAndBlobResultsIo.map { filesAndBlobResults =>
+      filesAndBlobResults map {
+        case (referenceFile, blobGetResult) =>
+          val crc32cFromManifest = BaseEncoding.base64.encode(
+            // drop 4 leading bytes from Long crc32c value
+            // https://stackoverflow.com/a/25111119/1794750
+            util.Arrays.copyOfRange(Longs.toByteArray(referenceFile.crc32c), 4, 8)
+          )
+
+          (referenceFile, crc32cFromManifest === blobGetResult.get().getCrc32c)
+      }
+    }
+  }
+
+  private def getMapOfValidReferenceFilePathsToDisks(gcsClient: Storage, manifestFile: ManifestFile): IO[Map[String, PipelinesApiReferenceFilesDisk]] = {
+    val refDisk = PipelinesApiReferenceFilesDisk(manifestFile.imageIdentifier, manifestFile.diskSizeGb)
+    val allReferenceFilesFromManifestMap = manifestFile.files.map(refFile => (refFile, refDisk)).toMap
+    val referenceFilesWithValidPathsIo = getReferenceFileToValidatedGcsPathMap(allReferenceFilesFromManifestMap.keySet)
+    val filesWithValidatedCrc32csIo = bulkValidateCrc32cs(gcsClient, referenceFilesWithValidPathsIo)
+
+    val validReferenceFilesFromManifestMapIo = filesWithValidatedCrc32csIo map {
+      filesWithValidatedCrc32c => {
+        allReferenceFilesFromManifestMap.filterKeys(key => filesWithValidatedCrc32c.getOrElse(key, false))
+      }
+    }
```

Comment:
TOL could the valid / invalid groupings be generated more easily with `partition`?